### PR TITLE
[WIP] no-run-benchmarks should imply building benchmarks

### PR DESCRIPTION
**Pull request by rabbitonweb**
Co-authored-by: Łukasz Gołębiewski <lukasz.golebiewski@gmail.com>

This PR will fix issue #4045. Work in progress.

Please include the following checklist in your PR:

* [ ] Any changes that could be relevant to users have been recorded in the ChangeLog.md
* [ ] The documentation has been updated, if necessary.

Integration test added under `test/integration/tests/4045-no-run-benchmarks/Main.hs`

### CODE REVIEW

File: test/integration/tests/1198-multiple-exes-with-same-name/Main.hs [hunk 1]
```diff
@@ -8,7 +8,7 @@ main = do
     stack [defaultResolverArg, "init", "--force"]
     stackCheckStderr
         ["build", "also-has-exe-foo", "has-exe-foo"]
-        (expectMessage buildMessage1)
+        (expectMessage buildMessage1)
```
Comment (mihaimaruseac) [inline test/integration/tests/1198-multiple-exes-with-same-name/Main.hs]: There is no need to add an extra whitespace at the end of the line here, can you please remove it?

File: test/integration/tests/4045-no-run-benchmarks/Main.hs [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+    stackFail ["bench"] 
+    stackFail ["test", "--no-run-benchmarks"] 
+    putStrLn "DONE"
+
+stackOk :: [String] -> IO ()
+stackOk arg = stackCheckStderr args (\_ -> return ())
+  
+stackFail :: [String] -> IO ()
+stackFail args = stackErrStderr args (\_ -> return ())
+
```
Comment (mihaimaruseac) [inline test/integration/tests/4045-no-run-benchmarks/Main.hs]: Remove the extra two empty lines at the end of this file

File: test/integration/tests/4045-no-run-benchmarks/files/.gitignore [hunk 3]
```diff
@@ -0,0 +1,3 @@
+.stack-work/
```
Comment (mihaimaruseac) [inline test/integration/tests/4045-no-run-benchmarks/files/.gitignore]: Remove the .gitignore file from the repo, there is a global .gitignore which already ignores the needed paths.

File: test/integration/tests/4045-no-run-benchmarks/files/ChangeLog.md [hunk 4]
```diff
@@ -0,0 +1,3 @@
+# Changelog for foff-bench-not-build
```
Comment (mihaimaruseac) [inline test/integration/tests/4045-no-run-benchmarks/files/ChangeLog.md]: We don't need this file in the repo, for just a test. Same for the LICENSE file and the readme

File: test/integration/tests/4045-no-run-benchmarks/Main.hs [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+import StackTest
+
+main :: IO ()
+main = do
+    stack [defaultResolverArg, "clean"]
+    stack [defaultResolverArg, "build"]
+    stackOk ["test"] 
+    stackFail ["bench"] 
+    stackFail ["test", "--no-run-benchmarks"] 
+    putStrLn "DONE"
```
Comment (mihaimaruseac) [inline test/integration/tests/4045-no-run-benchmarks/Main.hs]: I don't think we need the extra print in the test, probably is useful now while this is WIP though.

File: test/integration/tests/1198-multiple-exes-with-same-name/Main.hs (hunk 1 above)
Comment (rabbitonweb) [reply test/integration/tests/1198-multiple-exes-with-same-name/Main.hs]: 👍

File: test/integration/tests/4045-no-run-benchmarks/Main.hs (hunk 2 above)
Comment (rabbitonweb) [reply test/integration/tests/4045-no-run-benchmarks/Main.hs]: 👍

**Comment by rabbitonweb** [comment]
@mihaimaruseac this PR had a integration test that was not compiling (my fault). Interestingly the CI did notice that.

File: test/integration/tests/4045-no-run-benchmarks/files/.gitignore (hunk 3 above)
Comment (rabbitonweb) [reply test/integration/tests/4045-no-run-benchmarks/files/.gitignore]: 👍

File: test/integration/tests/4045-no-run-benchmarks/files/ChangeLog.md (hunk 4 above)
Comment (rabbitonweb) [reply test/integration/tests/4045-no-run-benchmarks/files/ChangeLog.md]: 👍

File: test/integration/tests/4045-no-run-benchmarks/Main.hs (hunk 5 above)
Comment (rabbitonweb) [reply test/integration/tests/4045-no-run-benchmarks/Main.hs]: 👍

File: test/integration/tests/4045-no-run-benchmarks/files/foff-bench-not-build.cabal [hunk 6]
```diff
@@ -0,0 +1,72 @@
+-- This file has been generated from package.yaml by hpack version 0.28.2.
```
Comment (mihaimaruseac) [inline test/integration/tests/4045-no-run-benchmarks/files/foff-bench-not-build.cabal]: Let's keep the cabal file in the repo, without superfluous comments and delete the hpack file. This ensures consistency with the other test suites.

**Comment by mihaimaruseac** [comment]
@rabbitonweb any more progress on this?

**Comment by lukasz-golebiewski** [comment]
We've made some progress, but ran out of ideas how this could be fixed. Could you give us a hint @mihaimaruseac ?

**Comment by mihaimaruseac** [comment]
Sorry, was away until now. I cannot look at this now, but will take a look tomorrow. Sorry for the delay

**Comment by mihaimaruseac** [comment]
So, if i understand #4045 right, all we need is to change the argument parser to add the `--bench` flag if `--no-run-benchmarks` was given (and similar for the tests).

In this way, the changes are isolated to just a single file, instead of being spread throughout the code.

Let me know if this makes sense

**Comment by mihaimaruseac** [comment]
Ping on this, as it's becoming stale.

**Comment by rabbitonweb** [comment]
please give time to the end of the weekend

**Comment by dbaynard** [comment]
@rabbitonweb Happy for you to continue with this if you still intend to do so — if you don't then please indicate and we'll try to find somebody to get it finished.

Thanks!

**Comment by snoyberg** [comment]
Haven't heard back from @rabbitonweb on this, closing for now.

**Comment by rabbitonweb** [comment]
Guys sorry I was not responding. I will reattempt and come back when I feel I have something working

**Pull request closed without merging**